### ISSUE TITLE
core: Write to files only after checking that Diff._file == Current state of file

### ISSUE THREAD

**Issue by meetmangukiya** [coala/coala#3912]
What coala does:
1. Loads all the files needed(passed from `-f`) into memory.
2. Bears propose results and hence create the diffs.
3. If there are no merge conflicts, by the end of the execution, the file is overwritten.

The problem arises when user changes file between calling coala i.e. loading and writing.

For example:
1. Suppose I had some error reported by PEP8Bear on a python file. It even proposed a patch. But since it was taking some time for execution, I switch windows and edited that file in my text editor. Now the file loaded in memory by coala and the file currently on harddrive are different. 
2. I switch back to coala console, and applied the patch.
3. What happens is: All the changes that I done, are lost! That is so not what I expect from `coala`.!

What I propose as a solution is:
1. Just before writing of the file, we read the file again and then check it against `Diff._file`. This is a performance drop, but the user data is safe!

**Comment by Makman2** [comment]
good idea. But don't read the file again, rather compare timestamps which is cheaper than reading it again :+1:

**Comment by meetmangukiya** [comment]
@Makman2 mark difficulty label if you are aware of this stuff.. :P 

Also, there are two facets to this issue: 
- Timestamp has not changed:
  - well this is simple, nothing has to be changed, just apply the patch as before :3 
- Timestamp has changed:
  - Now, since the timestamp has changed, the `Diff` has changed, now the important and sane thing to carry out will be compute a new `Diff` with the changed file and the changed introduced in the proposed `Diff`.
  - After this two things can be done, cross check the timestamp again and if changed again, compute `Diff` again. This can be a huge performance bottleneck and also can lead to stuck UI. I recommend we only check once.

Few things that I'd like to know is:
1. Does coala change timestamp of a file when it writes to it?
2. Does coala store timestamps of the loaded files(I guess no) ?

After these questions are answered, maybe I can get going ;)

**Comment by Siddhartha1234** [comment]
@meetmangukiya Coala uses timestamp for caching purposes . You can utilise that :)

**Comment by Makman2** [comment]
>   Now, since the timestamp has changed, the Diff has changed, now the important and sane thing to carry out will be compute a new Diff with the changed file and the changed introduced in the proposed Diff.

There might be conflicts, in this case we need to error out properly if a Diff reconstruction is not possible. Though on first implementation I would always emit an error that the file has changed and coala has to be run again.

> After this two things can be done, cross check the timestamp again and if changed again, compute Diff again. This can be a huge performance bottleneck and also can lead to stuck UI. I recommend we only check once.

Actually we can just use a write lock to prevent changes from outside during the 1st time Then we don't need to do this step again.

**Comment by meetmangukiya** [comment]
> Actually we can just use a write lock to prevent changes from outside during the 1st time Then we don't need to do this step again.

If something like this is possible, why don't we lock all the files that coala loaded, after loading ?

**Comment by Makman2** [comment]
Possible, though not sure how this impacts performance and how OS handles that^^

**Comment by meetmangukiya** [comment]
> A lock is a way to tell other threads or processes that a resource (like a file) is in use.

Seems like we should be having this already.

Some googling shows that the locking a file doesn't really make it read-only for that process. Also I looked at locking files in python, but they are mostly used to ensure the threads in that program don't write simultaneously. 

Maybe we can change file permissions to read only until coala executes. Most probably it will not be possible I guess.

**Comment by Makman2** [comment]
No I don't mean to use thread locks, but file lock mechanisms. For example just opening a file for write locks it for other apps for writing.

**Comment by meetmangukiya** [comment]
Blocked because no reliable mechanism/implementation method found yet..

**Comment by mrtes** [comment]
To pick this up again:
I haven't found a method to lock the file for writing either. 
I guess using the timestamp (and maybe also the size?) of the file is really the most sensible approach to check if the file was modified outside of coala.
To resolve the issue we would need to reload  (and cache) the file, make a diff between the original and new files and then apply the changes of the bears.

This raises I question I have not yet fully understood:
When is the file written back? Once the action is chosen, or at the end of the process?

**Comment by Makman2** [comment]
I believe at the end of the process if I recall correctly... we aggregate patches selected by the user and then apply them at the end.

**Comment by mrtes** [comment]
So a change would involve:

- Getting the cache to the method where the files are saved.
- Check the timestamp of all files that have to be changed according to the diffs.
- If they deviate, read the file and recalculate the diffs.
- Error or ask for user input on conflicts.
- Check timestamps again, and error out on further changes.
- Write out the changes.


Of course performance could suffer horrendously, so we should parallelize, or provide a cli option that disables (or enables?) this behaviour and just force writes or force errors or something.

(I know I basically just recompiled the thread so far, but I think its good to summarize)

What do you think about implications and scope of this?

**Comment by Makman2** [comment]
Phew parallelization outside the core... Would like to avoid that.

But yeah that seems to be it :+1:
I think this is going to be complicated though:
> Error or ask for user input on conflicts.

When coala is about to finish, we don't have official communication paths to the user yet. Putting another result into the queue handling a patch fail may not work^^
